hconfig_utils/mapl3g_generalized_equality.F90

```diff
@@ -0,0 +1,108 @@
+module mapl3g_generalized_equality
+
+   use :: esmf, only: ESMF_KIND_R4, ESMF_KIND_R8, ESMF_KIND_I4, ESMF_KIND_I8
+   implicit none
+   private
+
+   public :: are_equal
+
+   interface are_equal
+      module procedure :: equals_i4_scalar
+      module procedure :: equals_i8_scalar
+      module procedure :: equals_r4_scalar
+      module procedure :: equals_r8_scalar
+      module procedure :: equals_l_scalar
+      module procedure :: equals_string
+      module procedure :: equals_i4_array
+      module procedure :: equals_i8_array
+      module procedure :: equals_r4_array
+      module procedure :: equals_r8_array
+      module procedure :: equals_l_array
+   end interface
+
+contains
+
+   logical function equals_i4_scalar(u, v) result(lval)
+      integer(kind=ESMF_KIND_I4), intent(in) :: u, v
+
+      lval = (u == v)
+
+   end function equals_i4_scalar
+
+   logical function equals_i8_scalar(u, v) result(lval)
+      integer(kind=ESMF_KIND_I8), intent(in) :: u, v
+
+      lval = (u == v)
+
+   end function equals_i8_scalar
+
+   logical function equals_r4_scalar(u, v) result(lval)
+      real(kind=ESMF_KIND_R4), intent(in) :: u, v
+
+      lval = (u == v)
+
+   end function equals_r4_scalar
+
+   logical function equals_r8_scalar(u, v) result(lval)
+      real(kind=ESMF_KIND_R8), intent(in) :: u, v
+
+      lval = (u == v)
+
+   end function equals_r8_scalar
+
+   logical function equals_l_scalar(u, v) result(lval)
+      logical, intent(in) :: u, v
+
+      lval = (u .eqv. v)
+
+   end function equals_l_scalar
```

Comment:
Not a complaint, but wouldn't it have been much easier to just pass `.eqv.`, `==`, and `all(==)` as tokens to your template rather than implementing such trivial procedures?

Reply:
I originally had a number of macros like EQUALS_ that were redefined for each subroutine. In order to reduce the number of macros that were redefined for each subroutine, I switched to this. I'll look at this in the morning, but I think that this could be simplified with IS_ARRAY and IS_LOGICAL macros change the relation type.

Reply:
Note that it is also possible that you can use `all()` even in the scalar case.  Might be confusing, but if it simplifies the other logic...   I think it will become a common pattern when Fortran has proper templates.

Reply:
I'll leave this to the refactor. I wasn't able to use `ALL()` with a scalar, but there is probably a way to do that. In refactor, I'll probably just use IS_ARRAY and IS_LOGICAL to set the relational operation.